File: py/session.py
import hashlib
import hmac
import random
from parameter import *

from diffiehellman import diffiehellman

# patch DH with non RFC prime to be used for handshake
if not 1 in diffiehellman.PRIMES:
    diffiehellman.PRIMES.update({1: {
        "prime": 0xffffffffffffffffc90fdaa22168c234c4c6628b80dc1cd129024e088a67cc74020bbea63b139b22514a08798e3404ddef9519b3cd3a431b302b0a6df25f14374fe1356d6d51c245e485b576625e7ec6f44c42e9a63a3620ffffffffffffffff,
        "generator": 2
    }})

try:
    import protocol.impl.keyexchange_pb2 as keyexchange
    import protocol.impl.authentication_pb2 as authentication
except:
    raise Exception(
        "PROTO stubs were not found or have been corrupted. Please regenerate from .proto files using process_proto.py")


'''

'''


class Session:

    def __init__(self):
        # Generate local keys`
        self._local_keys = diffiehellman.DiffieHellman(
            group=1, key_length=KEY_LENGTH)
        self._local_keys.generate_private_key()
        self._local_keys.generate_public_key()

    def _sendClientHelloRequest(self):
        diffiehellman_hello = keyexchange.LoginCryptoDiffieHellmanHello(
            **{'gc': self._local_keys.public_key.to_bytes(KEY_LENGTH, byteorder='big'),
               'server_keys_known': 1})

        request = keyexchange.ClientHello(
            **{'build_info': keyexchange.BuildInfo(**{'product': keyexchange.PRODUCT_PARTNER,
                                                      'platform': keyexchange.PLATFORM_LINUX_X86,
                                                      'version': SPOTIFY_API_VERSION}),
               'cryptosuites_supported': [keyexchange.CRYPTO_SUITE_SHANNON],
               'login_crypto_hello': keyexchange.LoginCryptoHelloUnion(**{'diffie_hellman': diffiehellman_hello}),
               'client_nonce': bytes([int(random.random() * 0xFF) for x in range(0, 0x10)]),
               'padding': bytes([0x1E]),
               'feature_set': keyexchange.FeatureSet(**{'autoupdate2': True})})
        request_data = request.SerializeToString()
        return self._connection.send_packet(b"\x00\x04",
                                            request_data)

    def _processAPHelloResponse(self, init_client_packet):
        prefix, size, init_server_packet = self._connection.recv_packet()
        response = keyexchange.APResponseMessage()
        response.ParseFromString(init_server_packet[4:])

        remote_key = response.challenge.login_crypto_challenge.diffie_hellman.gs
        self._local_keys.generate_shared_secret(int.from_bytes(remote_key,
                                                               byteorder='big'))
        mac_original = hmac.new(self._local_keys.shared_secret.to_bytes(KEY_LENGTH,
                                                                        byteorder='big'),
                                digestmod=hashlib.sha1)
        data = []
        for i in range(1, 6):
            mac = mac_original.copy()
            mac.update(init_client_packet + init_server_packet + bytes([i]))
            digest = mac.digest()
            data += digest

        mac = hmac.new(bytes(data[:0x14]),
                       digestmod=hashlib.sha1)
        mac.update(init_client_packet + init_server_packet)
        return (mac.digest(),
                bytes(data[0x14: 0x34]),
                bytes(data[0x34: 0x54]))

    """ Send handsheke challenge """

    def _sendClientHandshakeChallenge(self,
                                      challenge):
        diffie_hellman = keyexchange.LoginCryptoDiffieHellmanResponse(
            **{'hmac': challenge})
        crypto_response = keyexchange.LoginCryptoResponseUnion(
            **{'diffie_hellman': diffie_hellman})
        packet = keyexchange.ClientResponsePlaintext(**{'login_crypto_response': crypto_response,
                                                        'pow_response': {},
                                                        'crypto_response': {}})
        self._connection.send_packet(prefix=None,
                                     data=packet.SerializeToString())

    def connect(self, connection):
        self._connection = connection
        # self._connection.connect()
        init_client_packet = self._sendClientHelloRequest()
        challenge, send_key, recv_key = self._processAPHelloResponse(
            init_client_packet)
        self._sendClientHandshakeChallenge(challenge)
        self._connection.handshake_completed(send_key,
                                             recv_key)
        return self

    def authenticate(self):

        auth_request = authentication.ClientResponseEncrypted(**{'login_credentials': authentication.LoginCredentials(**{'username':      "d",
                                                                                                                         'typ':           authentication.AUTHENTICATION_USER_PASS,
                                                                                                                         'auth_data':     bytes("d", 'ascii')}),
                                                                 'system_info': authentication.SystemInfo(**{'cpu_family':                authentication.CPU_UNKNOWN,
                                                                                                             'os':                        authentication.OS_UNKNOWN,
                                                                                                             'system_information_string': INFORMATION_STRING,
                                                                                                             'device_id':                 DEVICE_ID}),
                                                                 'version_string': VERSION_STRING})
        # auth_request = authentication.ClientResponseEncrypted(
        #     **{'system_info': authentication.SystemInfo(**{'cpu_family': authentication.CPU_UNKNOWN,
        #                                                    'os': authentication.OS_UNKNOWN,
        #                                                    'system_information_string': INFORMATION_STRING,
        #                                                    'device_id': DEVICE_ID}),
        #        'version_string': VERSION_STRING})
        # auth_request.login_credentials.CopyFrom(login_credentials)

        packet = self._connection.send_packet(LOGIN_REQUEST_COMMAND,
                                              auth_request.SerializeToString())
        # Get response
        command, size, body = self._connection.recv_packet()
        if command == AUTH_SUCCESSFUL_COMMAND:
            auth_welcome = authentication.APWelcome()
            auth_welcome.ParseFromString(body)
            print(auth_welcome)
            return auth_welcome.reusable_auth_credentials_type, auth_welcome.reusable_auth_credentials
        elif command == AUTH_DECLINED_COMMAND:
            raise Exception('AUTH DECLINED. Code: %02X' % command)

        raise Exception('UNKNOWN AUTH CODE %02X' % command)

File: src/Session.cpp
#include "Session.h"
#include "MercuryManager.h"

using random_bytes_engine = std::independent_bits_engine<std::default_random_engine, CHAR_BIT, uint8_t>;

const char *deviceId = "352198fd329622137e14901634264e6f332e5422";
const char *informationString = "cspot";
const char *versionString = "cspot-0.1";

Session::Session()
{
    // Generates the public and priv key
    this->localKeys = std::make_unique<DiffieHellman>();
    this->shanConn = std::make_shared<ShannonConnection>();
}

void Session::connect(std::shared_ptr<PlainConnection> connection)
{
    this->conn = connection;
    auto helloPacket = this->sendClientHelloRequest();
    this->processAPHelloResponse(helloPacket);
}

std::vector<uint8_t> Session::authenticate(std::string username, std::string password)
{
    ClientResponseEncrypted authRequest = {};
    authRequest.login_credentials.username = (char *)(username.c_str());
    authRequest.login_credentials.auth_data = stringToPBBytes(password);
    authRequest.login_credentials.typ = AuthenticationType_AUTHENTICATION_USER_PASS;
    authRequest.system_info.cpu_family = CpuFamily_CPU_UNKNOWN;
    authRequest.system_info.os = Os_OS_UNKNOWN;
    authRequest.system_info.system_information_string = (char *)informationString;
    authRequest.system_info.device_id = (char *)deviceId;
    authRequest.version_string = (char *)versionString;

    auto data = encodePB(ClientResponseEncrypted_fields, &authRequest);

    // Send login request
    this->shanConn->sendPacket(LOGIN_REQUEST_COMMAND, data);
    free(authRequest.login_credentials.auth_data);

    auto packet = this->shanConn->recvPacket();
    switch (packet->command)
    {
    case AUTH_SUCCESSFUL_COMMAND:
    {
        printf("Authorization successful\n");

        // @TODO store the reusable credentials
        auto welcomePacket = decodePB<APWelcome>(APWelcome_fields, packet->data);
        return std::vector<uint8_t>({0x1}); // TODO: return actual reusable credentaials to be stored somewhere
        break;
    }
    case AUTH_DECLINED_COMMAND:
    {
        printf("Authorization declined\n");
        break;
    }
    default:
        printf("Unknown auth fail code %d\n", packet->command);
    }

    return std::vector<uint8_t>(0);
}

void Session::processAPHelloResponse(std::vector<uint8_t> &helloPacket)
{
    auto data = this->conn->recvPacket();

    // Decode the response
    auto skipSize = std::vector<uint8_t>(data.begin() + 4, data.end());
    auto res = decodePB<APResponseMessage>(APResponseMessage_fields, skipSize);

    // Compute the diffie hellman shared key based on the response
    auto diffieKey = std::vector<uint8_t>(res.challenge.login_crypto_challenge.diffie_hellman.gs, res.challenge.login_crypto_challenge.diffie_hellman.gs + KEY_SIZE);
    auto sharedKey = this->localKeys->computeSharedKey(diffieKey);

    // Init client packet + Init server packets are required for the hmac challenge
    data.insert(data.begin(), helloPacket.begin(), helloPacket.end());

    // Solve the hmac challenge
    auto resultData = std::vector<uint8_t>(0);
    for (int x = 1; x < 6; x++)
    {
        auto challengeVector = std::vector<uint8_t>(1);
        challengeVector[0] = x;

        challengeVector.insert(challengeVector.begin(), data.begin(), data.end());
        auto digest = SHA1HMAC(sharedKey, challengeVector);
        resultData.insert(resultData.end(), digest.begin(), digest.end());
    }

    auto lastVec = std::vector<uint8_t>(resultData.begin(), resultData.begin() + 0x14);

    // Digest generated!
    auto digest = SHA1HMAC(lastVec, data);

    ClientResponsePlaintext response = {};
    response.login_crypto_response.has_diffie_hellman = true;

    std::copy(digest.begin(),
              digest.end(),
              response.login_crypto_response.diffie_hellman.hmac);

    auto resultPacket = encodePB(ClientResponsePlaintext_fields, &response);
    auto emptyPrefix = std::vector<uint8_t>(0);

    this->conn->sendPrefixPacket(emptyPrefix, resultPacket);

    // Get send and receive keys
    auto sendKey = std::vector<uint8_t>(resultData.begin() + 0x14, resultData.begin() + 0x34);
    auto recvKey = std::vector<uint8_t>(resultData.begin() + 0x34, resultData.begin() + 0x54);

    // Init shanno-encrypted connection
    this->shanConn->wrapConnection(this->conn, sendKey, recvKey);
}

std::vector<uint8_t> Session::sendClientHelloRequest()
{
    // Prepare protobuf message
    ClientHello request = ClientHello_init_default;

    // Copy the public key into diffiehellman hello packet
    std::copy(this->localKeys->publicKey.begin(),
              this->localKeys->publicKey.end(),
              request.login_crypto_hello.diffie_hellman.gc);

    request.login_crypto_hello.diffie_hellman.server_keys_known = 1;
    request.build_info.product = Product_PRODUCT_PARTNER;
    request.build_info.platform = Platform_PLATFORM_LINUX_X86;
    request.build_info.version = SPOTIFY_VERSION;
    request.feature_set.autoupdate2 = true;
    request.cryptosuites_supported[0] = Cryptosuite_CRYPTO_SUITE_SHANNON;
    request.padding[0] = 0x1E;

    request.has_feature_set = true;
    request.login_crypto_hello.has_diffie_hellman = true;
    request.has_padding = true;
    request.has_feature_set = true;

    // Generate the random nonce
    random_bytes_engine rbe;
    std::vector<uint8_t> nonce(16);
    std::generate(begin(nonce), end(nonce), std::ref(rbe));
    std::copy(nonce.begin(), nonce.end(), request.client_nonce);

    auto vecData = encodePB(ClientHello_fields, &request);

    auto prefix = std::vector<uint8_t>({0x00, 0x04});
    return this->conn->sendPrefixPacket(prefix, vecData);
}
